lumicks/pylake/force_calibration/power_spectrum.py

```diff
@@ -156,3 +157,105 @@ def plot(self, **kwargs):
+    peak_cutoff: float
+        Indicates what value of the normalized spectrum is deemed abnormally high. Default is 20.0,
+        which corresponds to a chance of about 2 in 1E9 that a peak of that magnitude occurs
+        naturally in an exponential distribution with rate parameter = 1.0. The minimum is 15.0.
+    baseline: float
+        The baseline level a peak needs to drop down to. Lower means that exponentially more data is
+        considered to be part of the peak. The default is 1.0, and the range of baseline is [0.0,
+        peak_cutoff]. No fit or data smoothing is performed, the peak starts or ends at the first
+        data point that satisfies the criterion.
+
+    Returns
+    -------
+    frequency_ranges: list of tuples (f_start, f_stop)
+        f_start is the frequency where a peak starts, and f_end is the frequency where a peak ends.
+    frequency_indices: list of tuples (start_index, stop_index)
+        start_index is the index in the array power_spectrum.frequencies where a peak starts.
+        Similarly, stop_index is the index in the array where a peak stops. This list of tuples is
+        only returned when return_indices is True.
+
+    Raises
+    ------
+    ValueError
+        Raises a ValueError if the function is called on a PowerSpectrum object with blocking
+        applied. This function only works for PowerSpectrum objects without blocking or windowing.
+    """
+
+    def find_ranges(list_of_numbers):
+        """Place a range of consecutive numbers in a tuple as (start, stop), and return a list of
+        tuples"""
+        ranges = []
+        for _, chunk in groupby(enumerate(list_of_numbers), lambda x: x[1] - x[0]):
+            chunk = list(chunk)
+            ranges += [(chunk[0][1], chunk[-1][1])]
+        return ranges
+
+    if power_spectrum.num_points_per_block != 1:
+        raise ValueError(
+            "identify_peaks only works if the power spectrum is not blocked / averaged"
+        )
+    peak_cutoff = max(15.0, peak_cutoff)
+    baseline = np.clip(baseline, 0.0, peak_cutoff)
+
+    # Normalize the spectrum
+    flattened_spectrum = power_spectrum.power / model_fun(power_spectrum.frequency)
+    peaks = np.where(flattened_spectrum > peak_cutoff)[0]
+
+    if not peaks.size:  # Nothing to do here
+        return [], [] if return_indices else []
+
+    raw_exclusion_ranges = []
+    baseline_values = np.where(flattened_spectrum < baseline)[0]
+    # Expand the range to the left and right of the peak, until we hit the first point below the
+    # baseline
+    for peak_start, peak_end in find_ranges(peaks):
+        peak_start = np.amax(baseline_values[baseline_values <= peak_start])
+        peak_end = np.amin(baseline_values[baseline_values >= peak_end])
+
+        raw_exclusion_ranges += [[peak_start, peak_end]]
+
+    # Figure out if any of the ranges overlap, and if so, concatenate them
+    raw_exclusion_ranges = sorted(raw_exclusion_ranges)
```

Comment:
I like that idea!

Reply:
I've benchmarked 3 different solutions that I could come up with:

- Find list of tuples(float, float) for values above baseline and peak_cutoff, and convert to 2D numpy array
- Then check for each peak inside a baseline range, and if so, store the baseline range (possible doubles)
```
for peak in peak_ranges:
            where = np.where(np.logical_and(baseline_ranges[:,0] <= peak[0], peak[0] < baseline_ranges[:,1]))[0]
            if where.size:
                exclusion_ranges.append(np.squeeze(baseline_ranges[where,:]))
                print(baseline_ranges[where,:])
```
*2.68 s* ± 110 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

- Find list of tuples(float, float) for values above baseline and peak_cutoff, and convert to 2D numpy array
- Then loop over all baseline ranges and check for existence of at least one peak
- Add the baseline if it contains a peak
```
exclusion_ranges = []
        for baseline_range in baseline_ranges:
            if np.any(peak_ranges[:,0] >= baseline_range[0]) and np.any(peak_ranges[:,0] < baseline_range[1]):
                exclusion_ranges.append(baseline_range)
```
*20.5 s* ± 759 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

- Original implementation - baseline_values as a 1D numpy array, for values below baseline
- Use numpy `amax` and `amin` functions to find largest range where value > baseline
- Clean up possible doubles or touching ranges
```
baseline_values = np.where(flattened_spectrum < baseline)[0]
[...]
for peak_start, peak_end in find_consecutive_ranges(peaks):
        peak_start = np.amax(baseline_values[baseline_values <= peak_start])
        peak_end = np.amin(baseline_values[baseline_values >= peak_end])

        raw_exclusion_ranges += [[peak_start, peak_end]]
+ overlap cleanup
```
*647 ms* ± 30.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

Based on performance, I'm inclined to keep the original implementation, unless you see another method that is about as fast (or faster) and more readable?

Reply:
Interesting.

I think my mistake was in not realizing how many baseline sections there would be (meaning you end up slicing and iterating frequently). Ideally, we should indeed avoid that.

How about something like the following? I didn't benchmark it, but given that it's all `numpy` and we slice less, I reckon/hope it should be pretty performant. Might still be a little slower than yours, given that I traverse the data a bit more often.

```python
def grab_contiguous_ranges(mask):
    # We cap the entire spectrum to make sure that we can handle spectra that begin
    # or end above the baseline in a uniform manner with ones that don't.
    # The cast to int is needed so that taking the derivative of this can produce
    # negative values
    capped_mask = np.diff(np.hstack((0, mask, 0), dtype=int))
    ranges = np.nonzero(np.abs(capped_mask))[0].reshape((-1, 2))
    ranges[:, 1] -= 1  # To keep it compatible with your current implementation
    return ranges

baseline_mask = flattened_spectrum > baseline
peak_mask = flattened_spectrum > peak_cutoff

peak_ranges = grab_contiguous_ranges(peak_mask)

if not peak_ranges.size:
    return []

baseline_ranges = grab_contiguous_ranges(baseline_mask)

# Find start points of baseline sections (int because diff can be negative).
start_points = np.diff(np.hstack((0, baseline_mask), dtype=int)) > 0

# This allows us to look up which baseline range to grab
baseline_indices = np.cumsum(start_points) - 1

# Any point inside the peak range will do to identify the baseline range
exclusion_baseline_indices = [
    baseline_indices[peak_position] for peak_position in peak_ranges[:, 0]
]

# Only report unique ones
exclusion_ranges = baseline_ranges[np.unique(exclusion_baseline_indices)]
```

Reply:
I like the `numpy`-based function for getting the contiguous ranges! I could not get my head around a proper approach but this is perfect. 

I can't get the same behavior with this exact implementation though. I created some pytests with what I expect as an outcome, and the results quite different. I can't really get my head around how the `cumsum` works in all of this. Can you explain how it works?

Reply:
Ah great! Tests! Nice case coverage on `identify_peaks` btw! It's always nice when you have tests to immediately check against. I think the difference is whether you decide that the baseline value is baseline or not (I incorrectly did `>` before). Try changing the baseline mask from:

`baseline_mask = flattened_spectrum > baseline`

to:

`baseline_mask = flattened_spectrum >= baseline`

As for how the second section works:

First I compute both masks and get the contiguous regions from them.
Next, I make a mask where the start point of each baseline section is marked by a single `True` (interpreted as `1` when doing numeric computation).
I get this start mask by taking the derivative of the baseline mask and looking where it's positive.
Cumsum is just cumulative integration, so, if you consider what we just made (in my best ASCII art):

```python
identity    baseline 1   baseline 2    baseline 3
          <============><==========><==============>
startmask 100000000000001000000000001000000000000000
cumsum    111111111111112222222222223333333333333333
```

Subtract one from this value and you have an array that allows us to look up which baseline is or was last active. 

Two details that do matter:
- If the baseline already starts at the starting edge, its starting position wouldn't be found. For this reason, I add a zero at the leftmost edge. The `np.diff` drops that sample again, so that cancels out nicely.
- Note that since we're only ever looking up which baseline to grab at a peak, and `peak_cutoff > baseline` by verified definition, we don't actually need any special handling that deals with baselines that have ended. We'll never look this value up at a place where the `intensity < baseline` anyhow.

Reply:
Right! So this did the trick for my sleepy head:
```
identity    baseline 1   baseline 2    baseline 3
          <============><==========><==============>
startmask 100000000000001000000000001000000000000000
cumsum    111111111111112222222222223333333333333333
```
I totally forgot about the zeros in the `diff` and sort of mentally made it a `where` statement. Clear now, and works!
```
324 ms ± 27.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
-> And faster :)